contractcourt/chain_watcher.go

```diff
@@ -769,6 +725,115 @@ func (c *chainWatcher) handleKnownLocalState(
 	return true, nil
 }
 
+// handleKnownRemoteState checks whether the passed spend is a remote state
+// that is known to us (a revoked, current or pending state). If so we will act
+// on this state using the passed chainSet. If this is not a known remote
+// state, false is returned.
+func (c *chainWatcher) handleKnownRemoteState(
+	commitSpend *chainntnfs.SpendDetail, broadcastStateNum uint64,
+	chainSet *chainSet) (bool, error) {
+
+	// If the channel is recovered, we won't have any remote commit to
+	// check against, so imemdiately return.
+	if c.cfg.chanState.HasChanStatus(channeldb.ChanStatusRestored) {
+		return false, nil
+	}
+
+	commitTxBroadcast := commitSpend.SpendingTx
+	commitHash := commitTxBroadcast.TxHash()
+	spendHeight := uint32(commitSpend.SpendingHeight)
+
+	switch {
+	// If the spending transaction matches the current latest state, then
+	// they've initiated a unilateral close. So we'll trigger the
+	// unilateral close signal so subscribers can clean up the state as
+	// necessary.
+	case chainSet.remoteCommit.CommitTx.TxHash() == commitHash:
+		log.Infof("Remote party broadcast base set, "+
+			"commit_num=%v", chainSet.remoteStateNum)
+
+		chainSet.commitSet.ConfCommitKey = &RemoteHtlcSet
+		err := c.dispatchRemoteForceClose(
+			commitSpend, chainSet.remoteCommit,
+			chainSet.commitSet,
+			c.cfg.chanState.RemoteCurrentRevocation,
+		)
+		if err != nil {
+			return false, fmt.Errorf("unable to handle remote "+
+				"close for chan_point=%v: %v",
+				c.cfg.chanState.FundingOutpoint, err)
+		}
+
+		return true, nil
+
+	// We'll also handle the case of the remote party broadcasting
+	// their commitment transaction which is one height above ours.
+	// This case can arise when we initiate a state transition, but
+	// the remote party has a fail crash _after_ accepting the new
+	// state, but _before_ sending their signature to us.
+	case chainSet.remotePendingCommit != nil &&
+		chainSet.remotePendingCommit.CommitTx.TxHash() == commitHash:
+
+		log.Infof("Remote party broadcast pending set, "+
+			"commit_num=%v", chainSet.remoteStateNum+1)
+
+		chainSet.commitSet.ConfCommitKey = &RemotePendingHtlcSet
+		err := c.dispatchRemoteForceClose(
+			commitSpend, *chainSet.remotePendingCommit,
+			chainSet.commitSet,
+			c.cfg.chanState.RemoteNextRevocation,
+		)
+		if err != nil {
+			return false, fmt.Errorf("unable to handle remote "+
+				"close for chan_point=%v: %v",
+				c.cfg.chanState.FundingOutpoint, err)
+		}
+
+		return true, nil
+	}
+
+	// We check if we have a revoked state at this state num that matches
+	// the spend transaction.
+	retribution, err := lnwallet.NewBreachRetribution(
```

Comment:
Since we here in this method want to handle the cases where we didn't lose data (`handleKnown[..]State`), we should be okay to just check if we have a breach retribution at this state number, then compare the txid.

The reason I opted for not comparing the state number against the remote height without any txid check, as before, is that the state number doesn't tell us whether this is a local or remote close. 

Instead we now use the process of elimination. We first eliminate the easy, known cases. If that fails we attempt to derive our scripts at a "future" state, and finally we go into DLP. (the reason we want DLP last is that you can never know when to give up).

If we were to rely only on the state number here to identify a breach, we would risk getting into the position where we have lost data, and then mistakenly think that a "future local" state is a breach. This is something I ran into when writing the test case that triggers this bug, and I found it easier to get known cases out of the way before attempting any of the recovery cases.